### PR TITLE
Implement the data models

### LINKED ISSUE #8: Implement the data models
### Description

To create the data models using TypeORM based on the provided ER diagram, we need to design the database schema with appropriate entities and their relationships. Here's a description of the process:

1. Identify Entities: Analyze the ER diagram and identify the entities involved in the project. In this case, the entities are likely to include "Mentor", "Mentee" and "platform".

2. Create Entity Classes: Create TypeScript classes for each entity identified. For example, you would create a `Mentor` class with properties like `country`, `expertise`, and `mentoringStrategy`. Similarly, create classes for related entities such as `Country` and `Expertise`, if necessary. These classes will represent the data models in your application.

3. Define Entity Relationships: Examine the ER diagram to determine the relationships between the entities. For instance, if there is a relationship between `Mentor` and `Country`, you would specify the appropriate decorators or methods in the `Mentor` class to establish that relationship.

4. Annotate Entity Classes: Use TypeORM decorators to annotate the entity classes with metadata, such as table names, column definitions, and relationship configurations. For example, you would use decorators like `@Entity`, `@Column`, and `@ManyToOne` to specify the table name, columns, and relationships for each entity class.

5. Create Database Tables: With the entity classes and annotations in place, you can use TypeORM's migration functionality or schema synchronization to create the corresponding database tables based on the defined entities.

ER diagram: https://drive.google.com/file/d/11KMgdNu2mSAm0Ner8UsSPQpZJS8QNqYc/view

Resources:

https://typeorm.io/entities

### Acceptance Criteria

* Relevant models are created under `/src/models` path

### PR DESCRIPTION
**Pull request by kumuditha-udayanga**
## Purpose

The purpose of this PR is to fix #8 

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [ ] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/entity/email.entity.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    recipient: string
+
+    @Column({type: 'varchar', length: 255})
+    subject: string
+
+    @Column({type: 'varchar', length: 655})
+    content: string
+    // Should be ENUM
+    @Column()
+    state: string
```
Comment (anjula-sack) [inline src/entity/email.entity.ts:24]: Let's create an enum type

File: src/entity/mentee.entity.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Column()
+    mentor_id: bigint
+
+    @CreateDateColumn()
+    created_at: Date | undefined
+
+    @UpdateDateColumn()
+    updated_at: Date | undefined
+
```
Comment (anjula-sack) [inline src/entity/mentee.entity.ts:38]: mente, mentor relationship

```
 @ManyToOne(() => Mentor, mentor => mentor.mentees)
  mentor: Mentor;
```

File: src/entity/mentor.entity.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    @OneToOne(() => profileEntity)
+    @JoinColumn()
+    profile: profileEntity
+
+    @CreateDateColumn()
+    created_at: Date | undefined
+
+    @UpdateDateColumn()
+    updated_at: Date | undefined
+
```
Comment (anjula-sack) [inline src/entity/mentor.entity.ts:36]:
```
@OneToMany(() => Mentee, mentee => mentee.mentor)
  mentees: Mentee[];
```

File: src/entity/mentor.entity.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+@Entity("mentor")
+class Mentor {
+    @PrimaryGeneratedColumn('uuid')
+    id!: bigint
+    // Should be Enum
+    @Column()
+    state: string
+    // Should be Enum
+    @Column()
+    category: string
```
Comment (anjula-sack) [inline src/entity/mentor.entity.ts:19]: This will be a string

**Review by anjula-sack — CHANGES_REQUESTED** [review]

**Comment by anjula-sack** [comment]
can you add the program entity as well @kumuditha-udayanga

File: src/entity/mentee.entity.ts [hunk 5]
```diff
@@ -1,18 +1,20 @@
...
+import profileEntity from "./profile.entity";
 
 @Entity('mentee')
 class Mentee {
     @PrimaryGeneratedColumn()
     id!: bigint
 
-    @Column({type: "enum"})
+    @Column()
     state: string
```
Comment (anjula-sack) [inline src/entity/mentee.entity.ts:18]: This is a enum. PENDING, REJECTED, ACCEPTED

File: src/entity/mentor.entity.ts (hunk 4 above)
Comment (anjula-sack) [reply src/entity/mentor.entity.ts:22]: Now this has a relationship. one to many with category

File: src/entity/platform.entity.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    landing_page_url: string
+
+    @Column()
+    email_templates: JSON
+
+    @Column()
+    title: string
+
+    @Column()
+    category: string
```
Comment (anjula-sack) [inline src/entity/platform.entity.ts:33]:
```suggestion
```

File: src/enums/enums.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+export enum ProfileTypes {
+    DEFAULT = "default",
+    ADMIN = "admin"
+}
+
+export enum EmailTypes {
```
Comment (anjula-sack) [inline src/enums/enums.ts:6]:
```suggestion
export enum EmailStatusTypes {
```

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/entity/category.entity.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    id!: bigint
+
+    @Column({type: 'varchar', length: 255})
+    category: string
+
+    constructor(
+        category: string,
+    ) {
+        this.category = category;
+    }
```
Comment (anjula-sack) [inline src/entity/category.entity.ts:19]:
```suggestion
    }
    
     @BeforeInsert()
  @BeforeUpdate()
  updateTimestamps() {
    this.updated_at = new Date();
    if (!this.uuid) {
      this.created_at = new Date();
    }
  }

  @BeforeInsert()
  async generateUuid() {
    if (!this.uuid) {
      this.uuid = uuidv4();
    }
  }
```

File: src/entity/category.entity.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import {
+    Column,
+    Entity,
+    PrimaryGeneratedColumn
+} from "typeorm";
+
+@Entity("category")
+class Category {
+    @PrimaryGeneratedColumn('uuid')
+    id!: bigint
```
Comment (anjula-sack) [inline src/entity/category.entity.ts:10]:
```suggestion
    uuid!: bigint
    
     @Column({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
     created_at: Date | undefined;

     @Column({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
     updated_at: Date | undefined;
```

File: src/entity/email.entity.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    CreateDateColumn,
+    Entity,
+    PrimaryGeneratedColumn
+} from "typeorm";
+import {EmailStatusTypes} from "../enums/enums";
+
+@Entity("email")
+class Email {
+    @PrimaryGeneratedColumn('uuid')
+    id!: bigint
```
Comment (anjula-sack) [inline src/entity/email.entity.ts:12]:
```suggestion
    uuid!: bigint
```

File: src/entity/email.entity.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    subject: string
+
+    @Column({type: 'varchar', length: 655})
+    content: string
+
+    @Column({type: 'enum', enum: EmailStatusTypes})
+    state: EmailStatusTypes
+
+    @CreateDateColumn()
+    created_at: Date | undefined
```
Comment (anjula-sack) [inline src/entity/email.entity.ts:27]:
```suggestion
    @Column({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
     created_at: Date | undefined;
     
     @Column({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
     updated_at: Date | undefined;
```

File: src/entity/email.entity.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        recipient: string,
+        subject: string,
+        content: string,
+        state: EmailStatusTypes
+    ) {
+        this.recipient = recipient;
+        this.subject = subject;
+        this.content = content;
+        this.state = state;
+    }
```
Comment (anjula-sack) [inline src/entity/email.entity.ts:39]:
```suggestion
    }
    
   @BeforeInsert()
  @BeforeUpdate()
  updateTimestamps() {
    this.updated_at = new Date();
    if (!this.uuid) {
      this.created_at = new Date();
    }
  }
  
  @BeforeInsert()
  async generateUuid() {
    if (!this.uuid) {
      this.uuid = uuidv4();
    }
  }
```

File: src/entity/mentee.entity.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+} from "typeorm";
+import {MenteeApplication} from "../types";
+import Mentor from "./mentor.entity";
+import profileEntity from "./profile.entity";
+import {MenteeStateTypes} from "../enums/enums";
+
+@Entity('mentee')
+class Mentee {
+    @PrimaryGeneratedColumn()
+    id!: bigint
```
Comment (anjula-sack) [inline src/entity/mentee.entity.ts:19]:
```suggestion
    uuid!: bigint
```

File: src/entity/mentor.entity.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    UpdateDateColumn
+} from "typeorm";
+import profileEntity from "./profile.entity";
+import Mentee from "./mentee.entity";
+import Category from "./category.entity";
+
+@Entity("mentor")
+class Mentor {
+    @PrimaryGeneratedColumn('uuid')
+    id!: bigint
```
Comment (anjula-sack) [inline src/entity/mentor.entity.ts:18]:
```suggestion
    uuid!: bigint
```

File: src/entity/mentor.entity.ts [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    profile: profileEntity
+
+    @OneToMany(() => Mentee, mentee => mentee.mentor)
+    mentees: Mentee[];
+
+    @CreateDateColumn()
+    created_at: Date | undefined
+
+    @UpdateDateColumn()
+    updated_at: Date | undefined
```
Comment (anjula-sack) [inline src/entity/mentor.entity.ts:43]: Update this just like above,

refer to the existing profile entity

File: src/entity/platform.entity.ts [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    CreateDateColumn,
+    Entity,
+    PrimaryGeneratedColumn,
+    UpdateDateColumn
+} from "typeorm";
+
+@Entity("platform")
+class Platform {
+    @PrimaryGeneratedColumn('uuid')
+    id!: bigint
```
Comment (anjula-sack) [inline src/entity/platform.entity.ts:12]:
```suggestion
    uuid!: bigint
```

File: src/types.ts [hunk 17]
```diff
@@ -0,0 +1,16 @@
+interface Option {
+// todo: To be determined
+}
```
Comment (anjula-sack) [inline src/types.ts:3]:
```suggestion
}

```

**Review by anjula-sack — CHANGES_REQUESTED** [review]

**Comment by anjula-sack** [comment]
@kumuditha-udayanga can you fix the conflicts?

**Comment by kumuditha-udayanga** [comment]
Sure @anjula-sack, I'll resolve them as well.

**Comment by anjula-sack** [comment]
Hi @kumuditha-udayanga, any updates on this?

**Review by anjula-sack — APPROVED** [review]
Thank you! @kumuditha-udayanga

**Pull request merged**